src/Symfony/Component/Message/DependencyInjection/MessagePass.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Message\DependencyInjection;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\Compiler\PriorityTaggedServiceTrait;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class MessagePass implements CompilerPassInterface
+{
+    use PriorityTaggedServiceTrait;
+
+    private $messageHandlerResolverService;
+    private $handlerTag;
+    private $messageBusService;
+    private $middlewareTag;
+
+    public function __construct(string $messageBusService = 'message_bus', string $middlewareTag = 'message_middleware', string $messageHandlerResolverService = 'message.handler_resolver', string $handlerTag = 'message_handler')
+    {
+        $this->messageHandlerResolverService = $messageHandlerResolverService;
+        $this->handlerTag = $handlerTag;
+        $this->messageBusService = $messageBusService;
+        $this->middlewareTag = $middlewareTag;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->hasDefinition($this->messageBusService)) {
+            return;
+        }
+
+        if (!$middlewares = $this->findAndSortTaggedServices($this->middlewareTag, $container)) {
+            throw new RuntimeException(sprintf('You must tag at least one service as "%s" to use the "%s" service.', $this->middlewareTag, $this->messageBusService));
+        }
+
+        $busDefinition = $container->getDefinition($this->messageBusService);
+        $busDefinition->replaceArgument(0, $middlewares);
+
+        $handlerResolver = $container->getDefinition($this->messageHandlerResolverService);
+        $handlerResolver->replaceArgument(0, $this->findHandlers($container));
+    }
+
+    private function findHandlers(ContainerBuilder $container): array
+    {
+        $handlersByMessage = array();
+
+        foreach ($container->findTaggedServiceIds($this->handlerTag, true) as $serviceId => $tags) {
+            foreach ($tags as $tag) {
+                $reflection = new \ReflectionClass($container->getDefinition($serviceId)->getClass());
+
+                try {
+                    $method = $reflection->getMethod('__invoke');
+                } catch (\ReflectionException $e) {
```

Comment:
just use `hasMethod`?

Reply:
`getMethod` has to be called nonetheless. As we need the access to the method parameters. Do we really need to call 2 methods instead of just catching the exception in those rare cases.

Reply:
> Do we really need to call 2 methods instead of just catching the exception

For readability / less LoC, i'd say yes. No real issue i guess.. just looks like an unneeded micro-optim.

Reply:
Note that we should also support having the `handles` attribute on the tag. In case a handler is handling multiple commands/queries and/or the typehint is not present, we need to be able to specify the handled message.